File: src/lib/projection/simulation/postings.ts
import type {
	Account,
	IsoDate,
	Posting,
	PostingFrequency,
} from "../types/model";

import { addMonthsClamped, compareIsoDates, daysBetween } from "../utils/date";
import {
	getHeadroom,
	getTotalDestinationHeadroom,
	getWithdrawableAmount,
} from "./accounts";
import { evaluateArithmetic } from "./arithmetic";

export interface DatedPostingOccurrence {
	posting: Posting;
	index: number;
}

export interface AccountMovementAction {
	sourceAccountId: string | null;
	destinations: string[] | null;
	requestedAmount: number;
	limitRemaining?: number;
}

export interface AccountMovementResult {
	requestedAmount: number;
	realizedAmount: number;
}

export function frequencyDivisor(frequency: PostingFrequency): number {
	switch (frequency) {
		case "daily":
			return 365;
		case "weekly":
			return 52;
		case "monthly":
			return 12;
		case "quarterly":
			return 4;
		case "annual":
			return 1;
	}
}

function advanceDate(
	date: IsoDate,
	frequency: PostingFrequency,
	periodCount: number,
): IsoDate {
	switch (frequency) {
		case "daily":
		case "weekly": {
			const daysPerPeriod = frequency === "daily" ? 1 : 7;
			const msPerDay = 24 * 60 * 60 * 1000;
			const base = new Date(`${date}T00:00:00Z`).getTime();
			const next = new Date(base + daysPerPeriod * periodCount * msPerDay);
			return next.toISOString().slice(0, 10);
		}
		case "monthly":
			return addMonthsClamped(date, periodCount);
		case "quarterly":
			return addMonthsClamped(date, periodCount * 3);
		case "annual":
			return addMonthsClamped(date, periodCount * 12);
	}
}

export function addOccurrences(
	postings: Posting[],
	eventDates: Map<IsoDate, DatedPostingOccurrence[]>,
	projectionStartDate: IsoDate,
	projectionEndDate: IsoDate,
	includeStartDate: boolean,
): void {
	postings.forEach((posting, index) => {
		if (!posting.enabled) {
			return;
		}

		const effectiveEndDate =
			posting.endDate !== null &&
			compareIsoDates(posting.endDate, projectionEndDate) < 0
				? posting.endDate
				: projectionEndDate;

		for (let periodCount = 0; ; periodCount += 1) {
			const occurrenceDate = advanceDate(
				posting.startDate,
				posting.frequency,
				periodCount,
			);
			if (compareIsoDates(occurrenceDate, effectiveEndDate) > 0) {
				break;
			}

			const startsInWindow = includeStartDate
				? compareIsoDates(occurrenceDate, projectionStartDate) >= 0
				: compareIsoDates(occurrenceDate, projectionStartDate) > 0;

			if (!startsInWindow) {
				continue;
			}

			const occurrences = eventDates.get(occurrenceDate) ?? [];
			occurrences.push({ posting, index });
			eventDates.set(occurrenceDate, occurrences);
		}
	});
}

export function applyAnnualGrowth(
	amount: number,
	annualGrowthRate: number,
	daysElapsed: number,
): number {
	if (amount === 0 || annualGrowthRate === 0 || daysElapsed <= 0) {
		return amount;
	}

	return amount * (1 + annualGrowthRate) ** (daysElapsed / 365);
}

export function computeRequestedAmount(
	occurrence: DatedPostingOccurrence,
	currentDate: IsoDate,
	latestRealizedPostingAmountById: Map<string, number>,
	balances: Record<string, number>,
	stochasticRate?: number,
): number {
	const { posting } = occurrence;

	const daysElapsed = daysBetween(posting.startDate, currentDate);
	const effectiveAnnualRate =
		stochasticRate !== undefined ? stochasticRate : posting.annualRate;
	const ratePerOccurrence =
		posting.annualRate === 0
			? 0
			: effectiveAnnualRate / frequencyDivisor(posting.frequency);

	const rawAmount = evaluateArithmetic(posting.arithmetic, {
		postingAmounts: latestRealizedPostingAmountById,
		accountBalances: balances,
		rate: ratePerOccurrence,
	});

	return applyAnnualGrowth(rawAmount, posting.annualGrowthRate, daysElapsed);
}

export function resolvePostingAmount(
	posting: Posting,
	requestedAmount: number,
	annualCapRemaining: number,
	balances: Record<string, number>,
	accountById: Map<string, Account>,
): number {
	return resolvePostingMovement(
		posting,
		requestedAmount,
		annualCapRemaining,
		balances,
		accountById,
	).realizedAmount;
}

export function resolvePostingMovement(
	posting: Posting,
	requestedAmount: number,
	annualCapRemaining: number,
	balances: Record<string, number>,
	accountById: Map<string, Account>,
): AccountMovementResult {
	return resolveAccountMovement(
		{
			sourceAccountId: posting.sourceAccountId,
			destinations: posting.destinations,
			requestedAmount,
			limitRemaining: annualCapRemaining,
		},
		balances,
		accountById,
	);
}

export function resolveAccountMovementAmount(
	action: AccountMovementAction,
	balances: Record<string, number>,
	accountById: Map<string, Account>,
): number {
	return resolveAccountMovement(action, balances, accountById).realizedAmount;
}

export function resolveAccountMovement(
	action: AccountMovementAction,
	balances: Record<string, number>,
	accountById: Map<string, Account>,
): AccountMovementResult {
	const requestedAmount = Math.max(0, action.requestedAmount);
	if (requestedAmount === 0) {
		return {
			requestedAmount,
			realizedAmount: 0,
		};
	}

	if (
		action.sourceAccountId !== null &&
		!accountById.has(action.sourceAccountId)
	) {
		return {
			requestedAmount,
			realizedAmount: 0,
		};
	}

	const sourceBalanceLimit =
		action.sourceAccountId === null
			? Number.POSITIVE_INFINITY
			: getWithdrawableAmount(balances, accountById, action.sourceAccountId);

	const destBalanceLimit =
		action.destinations === null
			? Number.POSITIVE_INFINITY
			: getTotalDestinationHeadroom(balances, accountById, action.destinations);

	const actionLimit = action.limitRemaining ?? Number.POSITIVE_INFINITY;
	const realizedAmount = Math.max(
		0,
		Math.min(
			requestedAmount,
			actionLimit,
			sourceBalanceLimit,
			destBalanceLimit,
		),
	);
	return {
		requestedAmount,
		realizedAmount,
	};
}

export function applyPosting(
	posting: Posting,
	realizedAmount: number,
	balances: Record<string, number>,
	accountById: Map<string, Account>,
): void {
	applyAccountMovement(
		{
			sourceAccountId: posting.sourceAccountId,
			destinations: posting.destinations,
			requestedAmount: realizedAmount,
		},
		realizedAmount,
		balances,
		accountById,
	);
}

export function applyAccountMovement(
	action: AccountMovementAction,
	realizedAmount: number,
	balances: Record<string, number>,
	accountById: Map<string, Account>,
): void {
	if (realizedAmount <= 0) {
		return;
	}

	if (action.sourceAccountId !== null) {
		balances[action.sourceAccountId] =
			(balances[action.sourceAccountId] ?? 0) - realizedAmount;
	}

	if (action.destinations === null) {
		return;
	}

	let remaining = realizedAmount;

	for (const destId of action.destinations) {
		if (remaining <= 0) {
			break;
		}

		const headroom = getHeadroom(balances, accountById, destId);
		if (headroom <= 0) {
			continue;
		}

		const allocated = Math.min(remaining, headroom);
		balances[destId] = (balances[destId] ?? 0) + allocated;
		remaining -= allocated;
	}
}

File: backend/internal/domain/postings.go
package domain

import (
	"math"

	"github.com/simhozebs/net-worth-estimator/backend/internal/types"
)

// Posting occurrence scheduling and movement resolution, ported from
// simulation/postings.ts.

// DatedPostingOccurrence binds a posting to its declaration index.
type DatedPostingOccurrence struct {
	Posting *types.Posting
	Index   int
}

// AccountMovementAction is a generic movement request against accounts.
type AccountMovementAction struct {
	SourceAccountID *string
	Destinations    []string
	RequestedAmount float64
	LimitRemaining  *float64
}

// AccountMovementResult reports requested vs realized amounts.
type AccountMovementResult struct {
	RequestedAmount float64
	RealizedAmount  float64
}

// FrequencyDivisor converts annual rates to per-occurrence rates.
func FrequencyDivisor(frequency types.PostingFrequency) int {
	switch frequency {
	case types.FrequencyOnce:
		return 1
	case types.FrequencyDaily:
		return 365
	case types.FrequencyWeekly:
		return 52
	case types.FrequencyMonthly:
		return 12
	case types.FrequencyQuarterly:
		return 4
	case types.FrequencyAnnual:
		return 1
	default:
		return 1
	}
}

func advanceDate(date string, frequency types.PostingFrequency, periodCount int) string {
	switch frequency {
	case types.FrequencyOnce:
		return date
	case types.FrequencyDaily, types.FrequencyWeekly:
		days := 7
		if frequency == types.FrequencyDaily {
			days = 1
		}
		t := MustParseIsoDate(date).AddDate(0, 0, days*periodCount)
		return FormatIsoDate(t)
	case types.FrequencyMonthly:
		return AddMonthsClamped(date, periodCount)
	case types.FrequencyQuarterly:
		return AddMonthsClamped(date, periodCount*3)
	case types.FrequencyAnnual:
		return AddMonthsClamped(date, periodCount*12)
	default:
		return date
	}
}

// AddOccurrences fills eventDates with occurrences in the projection window.
// Mirrors addOccurrences: once postings execute exactly on their start date;
// window inclusivity depends on includeStartDate.
func AddOccurrences(postings []types.Posting, eventDates map[string][]DatedPostingOccurrence, projectionStartDate, projectionEndDate string, includeStartDate bool) {
	for index := range postings {
		posting := &postings[index]
		if !posting.Enabled {
			continue
		}
		effectiveEndDate := projectionEndDate
		if posting.EndDate != nil && CompareIsoDates(*posting.EndDate, projectionEndDate) < 0 {
			effectiveEndDate = *posting.EndDate
		}
		for periodCount := 0; ; periodCount++ {
			if posting.Frequency == types.FrequencyOnce && periodCount > 0 {
				break
			}
			occurrenceDate := advanceDate(posting.StartDate, posting.Frequency, periodCount)
			if CompareIsoDates(occurrenceDate, effectiveEndDate) > 0 {
				break
			}
			var startsInWindow bool
			if includeStartDate {
				startsInWindow = CompareIsoDates(occurrenceDate, projectionStartDate) >= 0
			} else {
				startsInWindow = CompareIsoDates(occurrenceDate, projectionStartDate) > 0
			}
			if !startsInWindow {
				continue
			}
			eventDates[occurrenceDate] = append(eventDates[occurrenceDate], DatedPostingOccurrence{Posting: posting, Index: index})
		}
	}
}

// ApplyAnnualGrowth compounds an amount over elapsed days at an annual rate.
func ApplyAnnualGrowth(amount, annualGrowthRate float64, daysElapsed int) float64 {
	if amount == 0 || annualGrowthRate == 0 || daysElapsed <= 0 {
		return amount
	}
	return amount * math.Pow(1+annualGrowthRate, float64(daysElapsed)/365)
}

// ComputeRequestedAmount resolves the raw posting amount for one occurrence.
func ComputeRequestedAmount(occurrence DatedPostingOccurrence, currentDate string, latestRealized map[string]float64, realizedByYear map[string]map[string]float64, balances map[string]float64, stochasticRate *float64) (float64, error) {
	posting := occurrence.Posting
	daysElapsed := DaysBetween(posting.StartDate, currentDate)
	effectiveAnnualRate := posting.AnnualRate
	if stochasticRate != nil {
		effectiveAnnualRate = *stochasticRate
	}
	ratePerOccurrence := 0.0
	if posting.AnnualRate != 0 {
		ratePerOccurrence = effectiveAnnualRate / float64(FrequencyDivisor(posting.Frequency))
	}
	rawAmount, err := ResolvePostingAmountDescriptor(posting.Amount, &AmountProviderContext{
		Balances:                     balances,
		LatestRealizedPostingAmounts: latestRealized,
		RealizedPostingAmountsByYear: realizedByYear,
		Date:                         currentDate,
		OccurrenceRate:               ratePerOccurrence,
	})
	if err != nil {
		return 0, err
	}
	if posting.Amount.Resolver == "expression" {
		rawAmount = ApplyAnnualGrowth(rawAmount, posting.AnnualGrowthRate, daysElapsed)
	}
	return rawAmount, nil
}

// ResolveAccountMovement clamps a requested movement to constraints.
func ResolveAccountMovement(action AccountMovementAction, balances map[string]float64, accountByID map[string]types.Account) AccountMovementResult {
	requestedAmount := math.Max(0, action.RequestedAmount)
	if requestedAmount == 0 {
		return AccountMovementResult{RequestedAmount: requestedAmount, RealizedAmount: 0}
	}
	if action.SourceAccountID != nil && !accountExists(accountByID, *action.SourceAccountID) {
		return AccountMovementResult{RequestedAmount: requestedAmount, RealizedAmount: 0}
	}
	sourceBalanceLimit := math.Inf(1)
	if action.SourceAccountID != nil {
		sourceBalanceLimit = GetWithdrawableAmount(balances, accountByID, *action.SourceAccountID)
	}
	destBalanceLimit := math.Inf(1)
	if action.Destinations != nil {
		destBalanceLimit = GetTotalDestinationHeadroom(balances, accountByID, action.Destinations)
	}
	actionLimit := math.Inf(1)
	if action.LimitRemaining != nil {
		actionLimit = *action.LimitRemaining
	}
	realizedAmount := math.Max(0, math.Min(requestedAmount, math.Min(actionLimit, math.Min(sourceBalanceLimit, destBalanceLimit))))
	return AccountMovementResult{RequestedAmount: requestedAmount, RealizedAmount: realizedAmount}
}

// ResolvePostingMovement wraps ResolveAccountMovement for postings.
func ResolvePostingMovement(posting *types.Posting, requestedAmount, annualCapRemaining float64, balances map[string]float64, accountByID map[string]types.Account) AccountMovementResult {
	capRemaining := annualCapRemaining
	return ResolveAccountMovement(AccountMovementAction{
		SourceAccountID: posting.SourceAccountID,
		Destinations:    posting.Destinations,
		RequestedAmount: requestedAmount,
		LimitRemaining:  &capRemaining,
	}, balances, accountByID)
}

// ApplyAccountMovement applies a realized amount to balances.
func ApplyAccountMovement(action AccountMovementAction, realizedAmount float64, balances map[string]float64, accountByID map[string]types.Account) {
	if realizedAmount <= 0 {
		return
	}
	if action.SourceAccountID != nil {
		balances[*action.SourceAccountID] -= realizedAmount
	}
	if action.Destinations == nil {
		return
	}
	remaining := realizedAmount
	for _, destID := range action.Destinations {
		if remaining <= 0 {
			break
		}
		headroom := GetHeadroom(balances, accountByID, destID)
		if headroom <= 0 {
			continue
		}
		allocated := math.Min(remaining, headroom)
		balances[destID] += allocated
		remaining -= allocated
	}
}

// ApplyPosting applies a posting's realized movement.
func ApplyPosting(posting *types.Posting, realizedAmount float64, balances map[string]float64, accountByID map[string]types.Account) {
	ApplyAccountMovement(AccountMovementAction{
		SourceAccountID: posting.SourceAccountID,
		Destinations:    posting.Destinations,
		RequestedAmount: realizedAmount,
	}, realizedAmount, balances, accountByID)
}

func accountExists(accountByID map[string]types.Account, id string) bool {
	_, ok := accountByID[id]
	return ok
}
